src/qtism/common/datatypes/files/FileHash.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ *
+ * @author Julien Sébire <julien@taotesting.com>
+ * @license GPLv2
+ */
+
+namespace qtism\common\datatypes\files;
+
+use qtism\common\datatypes\QtiFile;
+use qtism\common\enums\BaseType;
+use qtism\common\enums\Cardinality;
+use RuntimeException;
+
+/**
+ * An implementation of File storing only a hash of the file.
+ * File contents has to be previously persisted externally.
+ * Hash string will serve the purpose of comparing files only.
+ */
+class FileHash implements QtiFile
+{
+    /**
+     * Key to use in json payload to trigger the storage of a file hash instead
+     * of a hash.
+     */
+    const FILE_HASH_KEY = 'fileHash';
+
+    /**
+     * The hash of the file contents.
+     *
+     * @var string
+     */
+    private $hash;
+
+    /**
+     * The MIME type of the file content.
+     *
+     * @var string
+     */
+    private $mimeType;
+
+    /**
+     * The path to the file on the external persistent storage.
+     *
+     * @var string
+     */
+    private $filename;
+
+    /**
+     * Create a new FileHash object.
+     *
+     * @param string $hash The hash of the file.
+     * @param string $mimeType The mime-type of the file.
+     * @param string $filename The path to the file on the external persistent storage.
+     * @throws RuntimeException If the file cannot be retrieved correctly.
+     */
+    public function __construct($hash, $mimeType, $filename)
+    {
+        $this->setFilename($filename);
+        $this->setMimeType($mimeType);
+        $this->setHash($hash);
+    }
+
+    /**
+     * Get the hash of the file.
+     *
+     * @return string
+     */
+    public function getHash()
+    {
+        return $this->hash;
+    }
+
+    /**
+     * Set the hash of the file.
+     *
+     * @param string $hash
+     */
+    protected function setHash($hash)
```

Comment:
Can we do hash validation here? Just to avoid receiving strings with any length.

Reply:
What kind of validation would that be? We don't have a limitation on the length on the SDK, and the length would vary depending on the hash function you use so I think a validation on the length should be made prior to sending the hash to the sdk.